File: naive_bayes/r/nb.r
# Define the DataPoint class
DataPoint <- setRefClass("DataPoint",
                         fields = list(features = "numeric",
                                       label = "numeric"))

# Define the Dataset class
Dataset <- setRefClass("Dataset",
                       fields = list(data = "list"))

# Function to load the dataset
loadDataset <- function() {
  # Load the dataset
  data <- list()

  # Your code to load the data from a file or any other source goes here
  # Append each data point as a DataPoint object to the data list

  dataset <- Dataset$new(data)
  return(dataset)
}

# Function to train the Naive Bayes classifier
trainNaiveBayes <- function(dataset) {
  numDataPoints <- length(dataset$data)
  numClasses <- 3  # Number of classes in your dataset
  numFeatures <- 4  # Number of features in your dataset

  classCounts <- rep(0, numClasses)
  priors <- rep(0, numClasses)
  likelihoods <- matrix(0, nrow = numClasses, ncol = numFeatures)

  # Count the occurrences of each class label
  for (i in 1:numDataPoints) {
    classCounts[[dataset$data[[i]]$label]] <- classCounts[[dataset$data[[i]]$label]] + 1
  }

  # Calculate priors
  for (i in 1:numClasses) {
    priors[i] <- classCounts[i] / numDataPoints
  }

  # Calculate likelihoods
  for (i in 1:numClasses) {
    for (j in 1:numFeatures) {
      featureSum <- 0.0
      featureCount <- 0

      # Sum the values of the feature for the current class
      for (k in 1:numDataPoints) {
        if (dataset$data[[k]]$label == i) {
          featureSum <- featureSum + dataset$data[[k]]$features[j]
          featureCount <- featureCount + 1
        }
      }

      # Calculate the average of the feature for the current class
      likelihoods[i, j] <- featureSum / featureCount
    }
  }

  return(list(priors = priors, likelihoods = likelihoods))
}

# Function to predict the class label for a new data point
predict <- function(dataPoint, priors, likelihoods) {
  numClasses <- length(priors)
  numFeatures <- ncol(likelihoods)
  maxPosterior <- 0.0
  predictedLabel <- -1

  # Calculate the posterior probability for each class
  for (i in 1:numClasses) {
    posterior <- priors[i]

    for (j in 1:numFeatures) {
      posterior <- posterior * exp(-(dataPoint$features[j] - likelihoods[i, j])^2 / 2)
    }

    # Update the predicted class if the posterior is higher than the current maximum
    if (posterior > maxPosterior) {
      maxPosterior <- posterior
      predictedLabel <- i
    }
  }

  return(predictedLabel)
}

# Main program
dataset <- loadDataset()
results <- trainNaiveBayes(dataset)

# Example usage: Predict the class label for a new data point
newDataPoint <- DataPoint$new(features = c(5.1, 3.5, 1.4, 0.2), label = 0)

predictedLabel <- predict(newDataPoint, results$priors, results$likelihoods)
cat("Predicted Label:", predictedLabel, "\n")

File: naive_bayes/cpp/nb.cpp
#include <iostream>
#include <vector>
#include <cmath>

#define NUM_CLASSES 3
#define NUM_FEATURES 4

struct DataPoint {
    std::vector<float> features;
    int label;
};

struct Dataset {
    std::vector<DataPoint> data;
};

Dataset loadDataset() {
    // Load the dataset
    Dataset dataset;

    // Your code to load the data from a file or any other source goes here
    // Assign the features and labels to the data vector in the dataset struct

    return dataset;
}

void trainNaiveBayes(const Dataset& dataset, std::vector<float>& priors, std::vector<std::vector<float>>& likelihoods) {
    int numDataPoints = dataset.data.size();
    std::vector<int> classCounts(NUM_CLASSES, 0);

    // Count the occurrences of each class label
    for (const auto& dataPoint : dataset.data) {
        classCounts[dataPoint.label]++;
    }

    // Calculate priors
    for (int i = 0; i < NUM_CLASSES; i++) {
        priors[i] = static_cast<float>(classCounts[i]) / numDataPoints;
    }

    // Calculate likelihoods
    for (int i = 0; i < NUM_CLASSES; i++) {
        for (int j = 0; j < NUM_FEATURES; j++) {
            float featureSum = 0.0f;
            int featureCount = 0;

            // Sum the values of the feature for the current class
            for (const auto& dataPoint : dataset.data) {
                if (dataPoint.label == i) {
                    featureSum += dataPoint.features[j];
                    featureCount++;
                }
            }

            // Calculate the average of the feature for the current class
            likelihoods[i][j] = featureSum / featureCount;
        }
    }
}

int predict(const DataPoint& dataPoint, const std::vector<float>& priors, const std::vector<std::vector<float>>& likelihoods) {
    float maxPosterior = 0.0f;
    int predictedClass = -1;

    // Calculate the posterior probability for each class
    for (int i = 0; i < NUM_CLASSES; i++) {
        float posterior = priors[i];

        for (int j = 0; j < NUM_FEATURES; j++) {
            posterior *= expf(-(powf(dataPoint.features[j] - likelihoods[i][j], 2.0f) / 2));
        }

        // Update the predicted class if the posterior is higher than the current maximum
        if (posterior > maxPosterior) {
            maxPosterior = posterior;
            predictedClass = i;
        }
    }

    return predictedClass;
}

int main() {
    Dataset dataset = loadDataset();
    std::vector<float> priors(NUM_CLASSES);
    std::vector<std::vector<float>> likelihoods(NUM_CLASSES, std::vector<float>(NUM_FEATURES));

    trainNaiveBayes(dataset, priors, likelihoods);

    // Example usage: Predict the class label for a new data point
    DataPoint newDataPoint;
    newDataPoint.features.push_back(5.1f);
    newDataPoint.features.push_back(3.5f);
    newDataPoint.features.push_back(1.4f);
    newDataPoint.features.push_back(0.2f);

    int predictedLabel = predict(newDataPoint, priors, likelihoods);
    std::cout << "Predicted Label: " << predictedLabel << std::endl;

    return 0;
}
